src/sentry/utils/performance_issues/performance_detection.py

```diff
@@ -1432,6 +1432,25 @@ def _maybe_performance_problem(self) -> Optional[PerformanceProblem]:
             offender_span_ids=[span["span_id"] for span in offender_spans],
         )
 
+    def _total_span_duration(self, spans: Sequence[Span]) -> timedelta:
+        """Returns the sum of time taken by spans, without double-counting overlaps."""
+        discrete_ranges = deque()
+        for span in spans:
+            start_time = timedelta(seconds=span.get("start_timestamp", 0))
+            end_time = timedelta(seconds=span.get("timestamp", 0))
+            if len(discrete_ranges) == 0:
+                discrete_ranges.append((start_time, end_time))
+            else:
+                last_start_time, last_end_time = discrete_ranges.pop()
+                if start_time < last_end_time:
+                    discrete_ranges.append((last_start_time, max(end_time, last_end_time)))
+                else:
+                    discrete_ranges.append((start_time, end_time))
+        return sum(
+            map(lambda range: range[1] - range[0], discrete_ranges),
+            timedelta(0),
+        )
```

Comment:
A much simpler implementation would have been to just return the difference between the end of the last span and the start of the first span. The difference is, that will include any gaps between spans. Sometimes those would be related (e.g. deserializing of results, etc), sometimes they would be unrelated work. The chosen approach is more conservative.

Reply:
Okay here, instead of 
```python
def sum_overlapping_intervals(a, b):
  sum, previous_end = a
  next_start, next_end = b
  if next_start < previous_end:
    return sum + next_end - previous_end, next_end
  return sum + next_end - next_start, next_end

def calculate_nonoverlap_intervals(l):
  return reduce(sum_overlapping_intervals, l, (0,0))[0]
```